lib/quantization/src/encoded_vectors_u8.rs

```diff
@@ -225,70 +260,183 @@ impl<TStorage: EncodedStorage> EncodedVectorsU8<TStorage> {
+
+                let score = match metadata.vector_parameters.distance_type {
+                    DistanceType::Dot | DistanceType::L2 => {
+                        impl_score_dot(q_ptr, v_ptr, metadata.actual_dim)
+                    }
+                    DistanceType::L1 => impl_score_l1(q_ptr, v_ptr, metadata.actual_dim),
+                };
 
-        let score = match self.metadata.vector_parameters.distance_type {
-            DistanceType::Dot | DistanceType::L2 => {
-                impl_score_dot(q_ptr, v_ptr, self.metadata.actual_dim)
+                metadata.multiplier * score as f32 + query.offset + vector_offset
             }
-            DistanceType::L1 => impl_score_l1(q_ptr, v_ptr, self.metadata.actual_dim),
-        };
+        }
+    }
 
-        self.metadata.multiplier * score as f32 + query.offset + vector_offset
+    pub fn score_point_simple_internal(&self, i: PointOffsetType, j: PointOffsetType) -> f32 {
+        match &self.metadata {
+            Metadata::Int8(metadata) => {
+                let (query_offset, q_ptr) = self.get_vec_ptr(i);
+                let (vector_offset, v_ptr) = self.get_vec_ptr(j);
+                let diff = metadata.get_internal_score_diff();
+
+                let score = match metadata.vector_parameters.distance_type {
+                    DistanceType::Dot | DistanceType::L2 => {
+                        impl_score_dot(q_ptr, v_ptr, metadata.actual_dim)
+                    }
+                    DistanceType::L1 => impl_score_l1(q_ptr, v_ptr, metadata.actual_dim),
+                };
+                metadata.multiplier * score as f32 + query_offset + vector_offset - diff
+            }
+        }
     }
 
     #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
     pub fn score_point_neon(&self, query: &EncodedQueryU8, bytes: &[u8]) -> f32 {
-        let (vector_offset, v_ptr) = Self::parse_vec_data(bytes);
-        let q_ptr = query.encoded_query.as_ptr();
+        match query {
+            EncodedQueryU8::Int8(query) => {
+                let metadata = match &self.metadata {
+                    Metadata::Int8(meta) => meta,
+                };
 
-        unsafe {
-            let score = match self.metadata.vector_parameters.distance_type {
-                DistanceType::Dot | DistanceType::L2 => {
-                    impl_score_dot_neon(q_ptr, v_ptr, self.metadata.actual_dim as u32)
-                }
-                DistanceType::L1 => {
-                    impl_score_l1_neon(q_ptr, v_ptr, self.metadata.actual_dim as u32)
+                let (vector_offset, v_ptr) = Self::parse_vec_data(bytes);
+                let q_ptr = query.encoded_query.as_ptr();
+
+                unsafe {
+                    let score = match metadata.vector_parameters.distance_type {
+                        DistanceType::Dot | DistanceType::L2 => {
+                            impl_score_dot_neon(q_ptr, v_ptr, metadata.actual_dim as u32)
```

Comment:
I think unsafe's scope should be limited to particular call (here and elsewhere).